### PR TITLE
rbd-mirror: replace remote pool polling with add/remove notifications

### PR DESCRIPTION
**Pull request by dillaman**

### CODE REVIEW

**Comment by trociny** [comment]
@dillaman observing crashes when running rbd_mirror(_stress).sh on teuthology:

http://pulpito.ceph.com/trociny-2016-12-09_06:43:36-rbd-wip-mgolub-testing---basic-mira/

and  locally:
```
2016-12-09 11:09:33.119818 7f783754cc40 -1 rbd::mirror::Mirror: 0x7f7840a47140 update_replayers: removing blacklisted replayer for uuid: 91c12cc7-d872-4cef-a12e-8b03c00b5aad cluster: cluster2 client: client.admin
2016-12-09 11:09:33.119843 7f783754cc40  5 rbd::mirror::PoolWatcher: 0x7f7840c6e830 shut_down: 
2016-12-09 11:09:33.119845 7f783754cc40  5 rbd::mirror::PoolWatcher: 0x7f7840c6e830 unregister_watcher: 
2016-12-09 11:09:33.119849 7f783754cc40  5 rbd::mirror::PoolWatcher: 0x7f7840c6e830 operator(): unregister_watcher: r=0
2016-12-09 11:09:33.120949 7f783754cc40 -1 /home/mgolub/ceph/ceph.upstream/src/librbd/Watcher.cc: In function 'virtual librbd::Watcher::~Watcher()' thread 7f783754cc40 time 2016-12-09 11:09:33.119857
/home/mgolub/ceph/ceph.upstream/src/librbd/Watcher.cc: 81: FAILED assert(m_watch_state != WATCH_STATE_REGISTERED)

 ceph version 11.0.2-2357-g9e1803f (9e1803fbaf6c82b47ac9ccf56a65a07fb4bcd8b6)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x82) [0x7f7837173a32]
 2: (librbd::Watcher::~Watcher()+0x337) [0x7f7837010937]
 3: (rbd::mirror::MirrorStatusWatchCtx::Watcher::~Watcher()+0x17) [0x7f7836f15947]
 4: (rbd::mirror::Replayer::~Replayer()+0x349) [0x7f7836f11b69]
 5: (std::_Rb_tree<std::pair<long, rbd::mirror::peer_t>, std::pair<std::pair<long, rbd::mirror::peer_t> const, std::unique_ptr<rbd::mirror::Replayer, std::default_delete<rbd::mirror::Replayer> > >, std::_Select1st<std::pair<std::pair<long, rbd::mirror::peer_t> const, std::unique_ptr<rbd::mirror::Replayer, std::default_delete<rbd::mirror::Replayer> > > >, std::less<std::pair<long, rbd::mirror::peer_t> >, std::allocator<std::pair<std::pair<long, rbd::mirror::peer_t> const, std::unique_ptr<rbd::mirror::Replayer, std::default_delete<rbd::mirror::Replayer> > > > >::_M_erase_aux(std::_Rb_tree_const_iterator<std::pair<std::pair<long, rbd::mirror::peer_t> const, std::unique_ptr<rbd::mirror::Replayer, std::default_delete<rbd::mirror::Replayer> > > >)+0x3b) [0x7f7836f0c1ab]
 6: (rbd::mirror::Mirror::update_replayers(std::map<long, std::set<rbd::mirror::peer_t, std::less<rbd::mirror::peer_t>, std::allocator<rbd::mirror::peer_t> >, std::less<long>, std::allocator<std::pair<long const, std::set<rbd::mirror::peer_t, std::less<rbd::mirror::peer_t>, std::allocator<rbd::mirror::peer_t> > > > > const&)+0x763) [0x7f7836f06253]
 7: (rbd::mirror::Mirror::run()+0xde) [0x7f7836f06dee]
 8: (main()+0x20e) [0x7f7836efbfde]
 9: (__libc_start_main()+0xf5) [0x7f782b946b45]
 10: (()+0x21e246) [0x7f7836f04246]
 NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.

2016-12-09 11:32:35.302769 7f4f80ff9700 -1 /home/mgolub/ceph/ceph.upstream/src/common/RWLock.h: In function 'void RWLock::get_read() const' thread 7f4f80ff9700 time 2016-12-09 11:32:35.301393
/home/mgolub/ceph/ceph.upstream/src/common/RWLock.h: 105: FAILED assert(r == 0)

 ceph version 11.0.2-2357-g9e1803f (9e1803fbaf6c82b47ac9ccf56a65a07fb4bcd8b6)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x82) [0x7f4fc1eb7a32]
 2: (()+0x2128c2) [0x7f4fc1c3c8c2]
 3: (librbd::ExclusiveLock<librbd::ImageCtx>::send_reacquire_lock()+0x4c8) [0x7f4fc1cd4738]
 4: (librbd::ExclusiveLock<librbd::ImageCtx>::reacquire_lock(Context*)+0x121) [0x7f4fc1cd56c1]
 5: (librbd::ImageWatcher<librbd::ImageCtx>::handle_rewatch_complete(int)+0x144) [0x7f4fc1cf6754]
 6: (librbd::Watcher::handle_rewatch(int)+0x437) [0x7f4fc1d55c97]
 7: (librbd::watcher::RewatchRequest::finish(int)+0x8d) [0x7f4fc1d57aed]
 8: (librbd::watcher::RewatchRequest::handle_rewatch(int)+0x105) [0x7f4fc1d58405]
 9: (librados::C_AioSafe::finish(int)+0x1d) [0x7f4fb8fa377d]
 10: (Context::complete(int)+0x9) [0x7f4fc1c597f9]
 11: (Finisher::finisher_thread_entry()+0x1f4) [0x7f4fc1eb6cd4]
 12: (()+0x80a4) [0x7f4fb8cf00a4]
 13: (clone()+0x6d) [0x7f4fb674f04d]
 NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.

```

**Comment by dillaman** [comment]
@trociny Pushed a fix for that crash

**Comment by trociny** [comment]
@dillaman looks like related:
```
[ RUN      ] TestLibRBD.FlushCacheWithCopyupOnExternalSnapshot
using new format!
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/librbd/test_librbd.cc:4853: Failure
      Expected: 0
To be equal to: rbd.clone(ioctx, name.c_str(), "one", ioctx, clone_name.c_str(), (1ULL<<0), &order)
      Which is: -2
[  FAILED  ] TestLibRBD.FlushCacheWithCopyupOnExternalSnapshot (10 ms)
```

**Comment by trociny** [comment]
@dillaman still observing crashes on teuthology:

http://pulpito.ceph.com/trociny-2016-12-23_09:28:57-rbd-wip-mgolub-testing---basic-smithi/
http://pulpito.ceph.com/trociny-2016-12-23_09:28:30-rbd-wip-mgolub-testing---basic-mira/
http://pulpito.ceph.com/trociny-2016-12-23_09:28:01-rbd-wip-mgolub-testing---basic-vps/

File: src/test/rbd_mirror/test_mock_PoolWatcher.cc [hunk 1]
```diff
@@ -0,0 +1,809 @@
...
+
+  void expect_work_queue(MockThreads &mock_threads) {
+    EXPECT_CALL(*mock_threads.work_queue, queue(_, _))
+      .WillRepeatedly(Invoke([this](Context *ctx, int r) {
+          m_threads->work_queue->queue(ctx, r);
+        }));
+  }
+
+  void expect_mirroring_watcher_is_unregister(MockMirroringWatcher &mock_mirroring_watcher,
+                                              bool unregistered) {
```
Comment (trociny) [inline src/test/rbd_mirror/test_mock_PoolWatcher.cc]: @dillaman May be the function name was supposed to be  `expect_mirroring_watcher_is_unregistered`?

**Comment by dillaman** [comment]
Note: there is a race condition with the ```ImageDeleter``` that cannot be fixed until #10896 is merged

**Comment by trociny** [comment]
Outdated

**Comment by trociny** [comment]
Sorry, wrong.

**Comment by trociny** [comment]
@dillaman May be a subset of this PR (namely, "utilize global image id as internal unique key" and "preliminary support to track multiple remote peer image sources") could be merged as a separate PR, not waiting for #10896?

I am just starting working on InstanceReplayerInterface [1] and it looks to me it would be good if your patches were merged before.

[1] http://tracker.ceph.com/issues/18785

File: src/tools/rbd_mirror/ImageReplayer.cc [hunk 2]
```diff
@@ -282,11 +274,11 @@ ImageReplayer<I>::ImageReplayer(Threads *threads,
   // re-registered using "remote_pool_name/remote_image_name" name.
 
   std::string pool_name;
-  int r = m_remote->pool_reverse_lookup(m_remote_pool_id, &pool_name);
+  int r = m_local->pool_reverse_lookup(m_local_pool_id, &pool_name);
   if (r < 0) {
-    derr << "error resolving remote pool " << m_remote_pool_id
+    derr << "error resolving remote pool " << m_local_pool_id
```
Comment (trociny) [inline src/tools/rbd_mirror/ImageReplayer.cc]: @dillaman I think this error message is confusing now -- I suppose users would thinking the ID in the message is from the remote cluster.

Comment (dillaman) [inline src/tools/rbd_mirror/ImageReplayer.cc]: Agreed -- probably should just pass in the pool name string since there really isn't a need to look it up.

**Comment by dillaman** [comment]
@trociny Sure -- I'll pull out the parts I can into a cleanup PR

File: src/tools/rbd_mirror/Replayer.cc [hunk 3]
```diff
@@ -770,36 +810,82 @@ bool Replayer::stop_image_replayer(unique_ptr<ImageReplayer<> > &image_replayer)
...
-    m_cond.Signal();
+    stop_image_replayers();
+
+    if (!m_image_replayers.empty()) {
+      Context *ctx = new FunctionContext([this, on_finish](int r) {
+          assert(r == 0);
+          stop_image_replayers(on_finish);
+        });
+      ctx = create_async_context_callback(m_threads->work_queue, ctx);
+      m_threads->timer->add_event_after(1, ctx);
```
Comment (trociny) [inline src/tools/rbd_mirror/Replayer.cc]: @dillaman Observing teuthology failures [1]

It looks like all cases are due to the timer lock is not held here:
```
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/12.0.0-1436-gb9f5564/rpm/el7/BUILD/ceph-12.0.0-1436-gb9f5564/src/common/Timer.cc: 127: FAILED assert(lock.is_locked())
 ceph version 12.0.0-1436-gb9f5564 (b9f556438d3c7612e9c0a042c8d8f0959cabd3a0)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x110) [0x7f351bd9c1a0]
 2: (()+0x2b840d) [0x7f351bd9440d]
 3: (rbd::mirror::Replayer::stop_image_replayers(Context*)+0x179) [0x7f3524d19119]
 4: (rbd::mirror::Replayer::handle_shut_down_pool_watcher(int, Context*)+0xb6) [0x7f3524d194a6]
 5: (FunctionContext::finish(int)+0x2a) [0x7f3524d1dc0a]
 6: (Context::complete(int)+0x9) [0x7f3524d1cc29]
 7: (ThreadPool::worker(ThreadPool::WorkThread*)+0xb59) [0x7f351bda4649]
 8: (ThreadPool::WorkThread::entry()+0x10) [0x7f351bda5660]
 9: (()+0x7dc5) [0x7f351a686dc5]
 10: (clone()+0x6d) [0x7f351914573d]
```
[1] http://pulpito.ceph.com/trociny-2017-03-16_07:53:00-rbd-wip-mgolub-testing---basic-smithi/

File: src/tools/rbd_mirror/PoolWatcher.cc [hunk 4]
```diff
@@ -23,123 +27,529 @@ using std::list;
...
+      decltype(Threads<I>::work_queue)>::type>::type;
+
+  MirroringWatcher(librados::IoCtx &io_ctx, ContextWQ *work_queue,
+                   PoolWatcher *pool_watcher)
+    : librbd::MirroringWatcher<I>(io_ctx, work_queue),
+      m_pool_watcher(pool_watcher) {
+  }
+
+  virtual void handle_rewatch_complete(int r) override {
+    m_pool_watcher->handle_rewatch_complete(r);
```
Comment (trociny) [inline src/tools/rbd_mirror/PoolWatcher.cc]: Is using both virtual and override intentional?

File: src/tools/rbd_mirror/PoolWatcher.cc [hunk 5]
```diff
@@ -23,123 +27,529 @@ using std::list;
...
+  }
+
+  virtual void handle_mode_updated(cls::rbd::MirrorMode mirror_mode) {
```
Comment (trociny) [inline src/tools/rbd_mirror/PoolWatcher.cc]: override

File: src/tools/rbd_mirror/PoolWatcher.cc [hunk 6]
```diff
@@ -23,123 +27,529 @@ using std::list;
...
+    // do nothing
+  }
+
+  virtual void handle_image_updated(cls::rbd::MirrorImageState state,
+                                    const std::string &remote_image_id,
+                                    const std::string &global_image_id) {
```
Comment (trociny) [inline src/tools/rbd_mirror/PoolWatcher.cc]: override

File: src/tools/rbd_mirror/PoolWatcher.cc [hunk 7]
```diff
@@ -23,123 +27,529 @@ using std::list;
...
+      ++it;
+      while (it != m_updated_images.end()) {
+        m_id_to_updated_images.erase({it->global_image_id,
+                                      it->remote_image_id});
+        it = m_updated_images.erase(it);
+      }
+    }
+    for (auto &updated_image : m_updated_images) {
+      updated_image.invalid = true;
+    }
```
Comment (trociny) [inline src/tools/rbd_mirror/PoolWatcher.cc]: Do we need this `for` loop, taking that after the code above we should have a single (invalid) in-flight request?

File: src/test/rbd_mirror/test_PoolWatcher.cc [hunk 8]
```diff
@@ -42,20 +44,48 @@ void register_test_pool_watcher() {
...
+  struct PoolWatcherListener : public PoolWatcher<>::Listener {
+    TestPoolWatcher *test;
+    Cond cond;
+    ImageIds image_ids;
+
+    PoolWatcherListener(TestPoolWatcher *test) : test(test) {
+    }
+
+    virtual void handle_update(const ImageIds &added_image_ids,
+                               const ImageIds &removed_image_ids) override {
```
Comment (trociny) [inline src/test/rbd_mirror/test_PoolWatcher.cc]: both virtual and override, it is intentional?

File: src/tools/rbd_mirror/Replayer.h [hunk 9]
```diff
@@ -58,19 +60,44 @@ class Replayer {
...
-  void init_local_mirroring_images();
-  void set_sources(const ImageIds &image_ids);
+  struct PoolWatcherListener : public PoolWatcher<>::Listener {
+    Replayer *replayer;
+
+    PoolWatcherListener(Replayer *replayer) : replayer(replayer) {
+    }
+
+    virtual void handle_update(const ImageIds &added_image_ids,
+                               const ImageIds &removed_image_ids) {
```
Comment (trociny) [inline src/tools/rbd_mirror/Replayer.h]: override

File: src/tools/rbd_mirror/PoolWatcher.cc (hunk 4 above)
Comment (dillaman) [inline src/tools/rbd_mirror/PoolWatcher.cc]: It's just very old code from before the switch-over -- I'll correct it

File: src/tools/rbd_mirror/PoolWatcher.cc (hunk 7 above)
Comment (dillaman) [inline src/tools/rbd_mirror/PoolWatcher.cc]: Agreed -- missed this during a refactor

**Comment by trociny** [comment]
@ceph-jenkins retest this please

**Comment by trociny** [comment]
@ceph-jenkins try again: retest this please

**Pull request merged**